File: example/src/test/java/ru/virgil/example/BuyingOrderApiTest.java
package ru.virgil.example;

import com.google.common.truth.Truth;
import lombok.RequiredArgsConstructor;
import org.apache.http.client.utils.URIBuilder;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.TestInstance;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.context.annotation.ComponentScan;
import ru.virgil.example.box.BoxController;
import ru.virgil.example.order.BuyingOrderDto;
import ru.virgil.example.order.BuyingOrderService;
import ru.virgil.example.truck.TruckDto;
import ru.virgil.example.user.UserDetails;
import ru.virgil.example.user.UserDetailsService;
import ru.virgil.example.util.security.user.WithMockFirebaseUser;
import ru.virgil.test_utils.fluent_request.RequestUtil;

import java.util.List;
import java.util.UUID;

import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;

@SpringBootTest
@AutoConfigureMockMvc
@WithMockFirebaseUser
@RequiredArgsConstructor(onConstructor_ = @Autowired)
@TestInstance(TestInstance.Lifecycle.PER_CLASS)
@ComponentScan("ru.virgil.test_utils")
public class BuyingOrderApiTest {

    public static final int PAGE = 0;
    public static final int PAGE_SIZE = 10;
    private final BuyingOrderService buyingOrderService;
    private final UserDetailsService userDetailsService;
    private final RequestUtil requestUtil;

    @Test
    void getAll() throws Exception {
        URIBuilder uriBuilder = new URIBuilder().setPath("/buying_order")
                .addParameter(BoxController.PAGE_PARAM, String.valueOf(PAGE))
                .addParameter(BoxController.PAGE_SIZE_PARAM, String.valueOf(PAGE_SIZE));
        List<BuyingOrderDto> buyingOrderDtoList = (List<BuyingOrderDto>) requestUtil.get(uriBuilder.toString())
                .receive(List.class, BuyingOrderDto.class)
                .and()
                .expect(status().isOk());
        Truth.assertThat(buyingOrderDtoList.stream().findAny().orElseThrow()).isInstanceOf(BuyingOrderDto.class);

        // todo: сначала обращаемся напрямую к серверу, а потом делаем запрос, не одновременно
        // todo: почему не проходит проверка?
        //        BuyingOrderDto randomBuyingOrderDto = requestUtil.get("/buying_order/%s".formatted(randomOrderUuid()))
        //                .receive(BuyingOrderDto.class)
        //                .expect(status().isOk());
        //        Truth.assertThat(buyingOrderDtoList).contains(randomBuyingOrderDto);

    }

    @Test
    void get() throws Exception {
        BuyingOrderDto randomBuyingOrderDto = (BuyingOrderDto) requestUtil.get("/buying_order/%s".formatted(randomOrderUuid()))
                .receive(BuyingOrderDto.class)
                .and()
                .expect(status().isOk());
        Truth.assertThat(randomBuyingOrderDto.getDescription()).isNotEmpty();
    }

    @Test
    void getTruckByOrder() throws Exception {
        String uri = new URIBuilder().setPathSegments("buying_order", randomOrderUuid().toString(), "truck")
                .addParameter(BoxController.PAGE_PARAM, String.valueOf(PAGE))
                .addParameter(BoxController.PAGE_SIZE_PARAM, String.valueOf(PAGE_SIZE))
                .toString();
        List<TruckDto> truckDtoList = (List<TruckDto>) requestUtil.get(uri)
                .receive(List.class, TruckDto.class)
                .and()
                .expect(status().isOk());
        Truth.assertThat(truckDtoList.stream().findAny().orElseThrow()).isInstanceOf(TruckDto.class);
    }

    private UUID randomOrderUuid() {
        UserDetails currentUser = userDetailsService.getCurrentUser();
        return buyingOrderService.getAll(currentUser, PAGE, PAGE_SIZE).stream()
                .findAny().orElseThrow()
                .getUuid();
    }
}

File: example/src/test/java/ru/virgil/example/BuyingOrderApiTest.kt
package ru.virgil.example

import com.google.common.truth.Truth
import org.apache.http.client.utils.URIBuilder
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.TestInstance
import org.springframework.beans.factory.ObjectProvider
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.beans.factory.annotation.Qualifier
import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.context.annotation.ComponentScan
import org.springframework.test.web.servlet.result.MockMvcResultMatchers
import ru.virgil.example.order.BuyingOrder
import ru.virgil.example.order.BuyingOrderDto
import ru.virgil.example.order.BuyingOrderMocker
import ru.virgil.example.system.rest.RestValues.PAGE_PARAM
import ru.virgil.example.system.rest.RestValues.PAGE_SIZE_PARAM
import ru.virgil.example.truck.TruckDto
import ru.virgil.example.util.security.user.WithMockFirebaseUser
import ru.virgil.test_utils.fluent_request.RequestUtil
import java.util.*

@SpringBootTest
@AutoConfigureMockMvc
@WithMockFirebaseUser
@TestInstance(TestInstance.Lifecycle.PER_CLASS)
@ComponentScan("ru.virgil.test_utils")
class BuyingOrderApiTest @Autowired constructor(
    private val requestUtil: RequestUtil,
    @Qualifier(BuyingOrderMocker.random)
    private val buyingOrderProvider: ObjectProvider<BuyingOrder>,
) {

    @Throws(Exception::class)
    @Test
    fun getAll(): Unit {
        val uriBuilder = URIBuilder().setPath("/buying_order")
            .addParameter(PAGE_PARAM, PAGE.toString())
            .addParameter(PAGE_SIZE_PARAM, PAGE_SIZE.toString())
        val buyingOrderDtoList = requestUtil.get(uriBuilder.toString())
            .receive(MutableList::class.java, BuyingOrderDto::class.java)
            .and()
            .expect(MockMvcResultMatchers.status().isOk) as List<BuyingOrderDto>
        Truth.assertThat(buyingOrderDtoList.stream().findAny().orElseThrow())
            .isInstanceOf(BuyingOrderDto::class.java)
    }

    @Test
    @Throws(Exception::class)
    fun get() {
        val buyingOrder = buyingOrderProvider.getObject()
        val randomBuyingOrderDto = requestUtil.get("/buying_order/${buyingOrder.uuid}")
            .receive(BuyingOrderDto::class.java)
            .and()
            .expect(MockMvcResultMatchers.status().isOk) as BuyingOrderDto
        Truth.assertThat(randomBuyingOrderDto.description).isNotEmpty()
    }

    @Throws(Exception::class)
    @Test
    fun getTruckByOrder(): Unit {
        val buyingOrder = buyingOrderProvider.getObject()
        val uri = URIBuilder().setPathSegments("buying_order", buyingOrder.uuid.toString(), "truck")
            .addParameter(PAGE_PARAM, PAGE.toString())
            .addParameter(PAGE_SIZE_PARAM, PAGE_SIZE.toString())
            .toString()
        val truckDtoList = requestUtil.get(uri)
            .receive(MutableList::class.java, TruckDto::class.java)
            .and()
            .expect(MockMvcResultMatchers.status().isOk) as List<TruckDto>
        Truth.assertThat(truckDtoList.stream().findAny().orElseThrow()).isInstanceOf(TruckDto::class.java)
    }

    companion object {

        const val PAGE = 0
        const val PAGE_SIZE = 10
    }
}
